TASK: Migrate to PHPStan for Flow 8

mhsdesign opened the pull request:
This is a backport of https://github.com/neos/flow-development-collection/pull/3216

Adds PHPStan level 1 to the whole Flow code base and CI.
Psalm was removed.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [x] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked wit `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Up phpstan to level 1 in Neos.Flow

mhsdesign committed:
TASK: Up phpstan to level 1 to all packages

mhsdesign committed:
TASK: Remove psalm from ci and run phpstan

mhsdesign committed:
TASK: `resource:clean` abort if `Neos.Media` is not installed

mhsdesign committed:
TASK: use `@phpstan-consistent-constructor` instead of `final` constructor

... as flows proxy building doesnt like it when on multiple level of inheritance:

Cannot override final method Neos\Flow\Security\Authentication\Provider\AbstractProvider::__construct() in /home/runner/work/flow-development-collection/flow-development-collection/flow-development-distribution/Data/Temporary/Testing/Cache/Code/Flow_Object_Classes/Neos_Flow_Security_Authentication_Provider_TestingProvider.php on line 130

see also https://phpstan.org/blog/solving-phpstan-error-unsafe-usage-of-new-static

mhsdesign committed:
TASK: Remove psalm annotations and baseline

mhsdesign committed:
TASK: Add phpstan level 0 to Neos.Flow

mhsdesign committed:
TASK: Ignore error in likely dead code path

mhsdesign commented:
We would need to adjust https://github.com/neos/flow-development-distribution to provide phpstan

mhsdesign commented on Neos.Eel/Classes/ProtectedContext.php:
```diff
@@ -67,31 +67,6 @@ public function getAndWrap($path = null)
-     * @return void
-     * @deprecated Use allow() instead. See https://github.com/neos/flow-development-collection/pull/2024
-     */
-    public function whitelist($pathOrMethods)
```
todo readd and drop with 9.0

mhsdesign commented on Neos.Flow/Classes/Http/Cookie.php:
```diff
@@ -246,7 +246,7 @@ public static function createFromRawSetCookieHeader($header)
         }
 
-        $cookie = new static(
+        $cookie = new self(
```
todo revert

mhsdesign commented on Neos.Flow/Classes/Mvc/ActionResponse.php:
```diff
@@ -144,30 +142,6 @@ public function deleteCookie(string $cookieName): void
-     * @return void
-     * @deprecated since Flow 7.0 use setHttpHeader or replaceHttpResponse instead. For now this will still work with $componentClassName of "SetHeaderComponent" or "ReplaceHttpResponseComponent" only.
-     */
-    public function setComponentParameter(string $componentClassName, string $parameterName, $value): void
```
todo readd and drop with 9.0

mhsdesign commented on Neos.Flow/Classes/Persistence/EmptyQueryResult.php:
```diff
@@ -66,92 +66,55 @@ public function toArray(): array
-     */
-    #[\ReturnTypeWillChange]
-    public function current()
+    public function current(): mixed
```
todo readd and drop with 9.0

mhsdesign commented on Neos.Flow/Classes/Reflection/Exception/ClassLoadingForReflectionFailedException.php:
```diff
@@ -20,12 +20,7 @@ class ClassLoadingForReflectionFailedException extends \Neos\Flow\Reflection\Exc
-        } else {
-            $message = sprintf('Required class "%s" could not be loaded properly for reflection from "%s".%3$s%3$sPossible reasons are:%3$s%3$s * Requiring non-existent classes%3$s * Using non-supported annotations%3$s * Class-/filename mismatch.%3$s%3$sThe "Neos.Flow.object.includeClasses" setting can be used to include or exclude classes from reflection.', $className, $reflectedClass, chr(10));
-        }
+        $message = sprintf('Required class "%s" could not be loaded properly for reflection from "%s".%3$s%3$sPossible reasons are:%3$s%3$s * Requiring non-existent classes%3$s * Using non-supported annotations%3$s * Class-/filename mismatch.%3$s%3$sThe "Neos.Flow.object.includeClasses" setting can be used to include or exclude classes from reflection.', $className, $reflectedClass, chr(10));
```
todo readd and drop with 9.0

mhsdesign committed:
TASK: Up phpstan to level 0 to all packages

mhsdesign commented on Neos.Flow/Classes/Security/Channel/HttpsInterceptor.php:
```diff
@@ -48,5 +49,6 @@ public function __construct(
      */
     public function invoke()
     {
+        return true;
```
Hello dear reviewer, are the changes here correct?

kdambekalns replied:
Well… I think this interceptor was never properly implemented. 😬 So it might even be totally irrelevant.

kitsunet replied:
Yeah seems not properly implemented, I am fine with this change for now, _maybe_ a TODO / FIXME ala "throw away or finish"

mhsdesign commented on Neos.Flow/Classes/Security/Context.php:
```diff
@@ -710,16 +710,10 @@ protected function isTokenActive(TokenInterface $token)
      * @return array Array of Authentication\TokenInterface objects
      */
-    protected function mergeTokens($managerTokens, $sessionTokens)
+    protected function mergeTokens(array $managerTokens, array $sessionTokens)
```
this is a 100% safe change, as the calling side promises to provide an array

mhsdesign commented on Neos.Flow/Classes/Security/Authorization/Interceptor/RequireAuthentication.php:
```diff
@@ -39,12 +39,13 @@ public function __construct(AuthenticationManagerInterface $authenticationManage
     public function invoke()
     {
         $this->authenticationManager->authenticate();
+        return true;
```
Dear reader. Are the changes here correct?

kdambekalns replied:
Yes, `authenticate()` throws, if it fails.

mhsdesign reviewed (commented):
.

kdambekalns commented on Neos.Flow/Classes/Aop/Builder/AdvicedConstructorInterceptorBuilder.php:
```diff
@@ -40,6 +40,11 @@ public function build(string $methodName, array $interceptedMethods, string $tar
         $proxyMethod = $this->compiler->getProxyClass($targetClassName)->getConstructor();
         if ($declaringClassName !== $targetClassName) {
+            /**
+             * @phpstan-ignore-next-line "buildMethodParametersCode" does not exist, and never has since the introduction of this call:
```
Interesting… the method does exist, but not in a way that easily explains why it would be accessible on `$this` at that point. 🤔 Maybe @robertlemke remembers more?

mhsdesign replied:
Well i just think this codepath is never taken ...

kitsunet replied:
This dates back to Andi and I believe the solution is as simple as this being 
`$proxyMethod->setMethodParametersCode($proxyMethod->buildMethodParametersCode($declaringClassName, $methodName, true));`

Which becomes even clearer looking at the sibling code in `AdvicedMethodInterceptorBuilder` where it is like I suggested

mhsdesign replied:
but it seems no one ever noticed this, and it never caused trouble? So its a dead code path?

either way i fixed it, cant be worse than a fatal error. :D

kdambekalns commented on Neos.Flow/Classes/Command/ResourceCommandController.php:
```diff
@@ -229,26 +229,28 @@ public function cleanCommand()
-                /* @var ThumbnailRepository $thumbnailRepository */
-                $thumbnailRepository = $this->objectManager->get(ThumbnailRepository::class);
-            }
+        if (!class_exists(AssetRepository::class) || !class_exists(ThumbnailRepository::class)) {
+            // @todo flow has no dependency on Neos.Media. This code should be extracted.
+            $this->outputLine('Neos.Media is not installed. Skipping cleanup of assest and thumbnails.');
+            $this->quit();
+        }
```
Not correct. This skips the cleanup completely, but that's not what code did so far.

mhsdesign replied:
upsi good catch yes adjusted as requested

kdambekalns commented on Neos.Flow/Classes/Mvc/Routing/IdentityRoutePart.php:
```diff
@@ -55,8 +55,7 @@ class IdentityRoutePart extends DynamicRoutePart
      *
-     * @var string
-     * @psalm-var class-string
+     * @var class-string
```
What does this do to reflection? Might not be important here, but e.g. in a entity this would probably be breaking.

mhsdesign replied:
yes in entities this will not work at all.

My rule is to use those modern annotations everywhere except in entities and controllers as flow will crash otherwise.

And in this case i like to leave it here.

kdambekalns commented on Neos.Kickstarter/Classes/Command/KickstartCommandController.php:
```diff
@@ -47,7 +47,6 @@ class KickstartCommandController extends CommandController
      *
      * @param string $packageKey The package key, for example "MyCompany.MyPackageName"
      * @param string $packageType Optional package type, e.g. "neos-plugin"
-     * @return string
```
Strictly speaking (as per our CGL) `@return` must be given, unless it's on `__construct` or a return type declaration is present. So why not add the latter (for all those places)?

mhsdesign replied:
hmm yes true, thats on my board for the later phpstan levels where it will also be enforced okay?

kdambekalns commented on Neos.Kickstarter/Classes/Utility/Inflector.php:
```diff
@@ -24,6 +24,7 @@ class Inflector
      */
     public function pluralize($word)
     {
+        /** @phpstan-ignore-next-line we could use autoloading instead */
```
That comment doesn't explain anything to me… how would autoloading help with a static method call?

mhsdesign replied:
the class is an ancient php library that we have in the resources folder and require_once via autoloading. Phpstan doesnt now any of it ^^

kitsunet replied:
Yep, instead of doing it in the Package.php we should just make this a https://getcomposer.org/doc/04-schema.md#files

mhsdesign replied:
will try it with flow 9

kdambekalns commented on Readme.rst:
```diff
@@ -69,8 +69,7 @@ To commit changes to the framework switch into the Framework directory (``cd Pac
 
-To run tests, run ``./bin/phpunit -c ./Build/BuildEssentials/PhpUnit/UnitTests.xml`` for unit or ``./bin/phpunit -c ./Build/BuildEssentials/PhpUnit/FunctionalTests.xml`` for functional/integration tests. If you are on 6.0 or later, you can also run ``./bin/psalm --config=Packages/Framework/psalm.xml``
-to run static analysis tools.
+To run tests, run ``./bin/phpunit -c ./Build/BuildEssentials/PhpUnit/UnitTests.xml`` for unit or ``./bin/phpunit -c ./Build/BuildEssentials/PhpUnit/FunctionalTests.xml`` for functional/integration tests and for static analysis ``composer lint``.
```
Those commands will not work "the same way", as for the`composer lint` call you need to `cd Packages/Framework` first, no? Or, if you are "there", the PhpUnit calls will not work, as `Build/BuildEssentials` doesn't exist.

kdambekalns reviewed (changes requested):
Mostly comments, but one real change request for the `ResourceCommandController`

kitsunet reviewed (approved):
I think I am fine with this overall for now, thanks for putting the work in!

mhsdesign committed:
TASK: Fix `AdvicedConstructorInterceptorBuilder`

The method `buildMethodParametersCode` does not exist and most likely `$proxyMethod->buildMethodParametersCode` is meant.

> This dates back to Andi and I believe the solution is as simple as this being
> $proxyMethod->setMethodParametersCode($proxyMethod->buildMethodParametersCode($declaringClassName, $methodName, true));

> Which becomes even clearer looking at the sibling code in AdvicedMethodInterceptorBuilder where it is like I suggested

https://github.com/neos/flow-development-collection/pull/3218#discussion_r1384602384

kdambekalns reviewed (approved):
Yeah, why not.

kdambekalns commented:
… but fix the PHPStan complaints first. 🤣

mhsdesign commented:
Thank you for your quick reviews ❤️
